src/main/java/uk/gov/companieshouse/itemhandler/kafka/OrdersKafkaConsumer.java

```diff
@@ -1,173 +1,256 @@
+     * the message is published to `-error` topic for failover processing.
+     * @param message
+     * @throws SerializationException
+     * @throws ExecutionException
+     * @throws InterruptedException
+     */
+    @KafkaListener(id = ORDER_RECEIVED_GROUP, groupId = ORDER_RECEIVED_GROUP,
+                    topics = ORDER_RECEIVED_TOPIC,
+                    autoStartup = "#{!${uk.gov.companieshouse.item-handler.error-consumer}}")
+    public void processOrderReceived(org.springframework.messaging.Message<OrderReceived> message)
+            throws SerializationException, ExecutionException, InterruptedException {
+        OrderReceived orderReceived = message.getPayload();
+        String orderReceivedUri = orderReceived.getOrderUri();
+        try {
+            logMessageReceived(message, ORDER_RECEIVED_TOPIC);
 
-        consumerConfig.setTopics(Collections.singletonList(ORDER_RECEIVED_TOPIC_RETRY));
-        consumerConfig.setGroupName(ORDER_RECEIVED_GROUP_RETRY);
-        chKafkaConsumerGroupRetry
-                = new CHKafkaResilientConsumerGroup(consumerConfig, CHConsumerType.RETRY_CONSUMER,
-                        new KafkaConsumerFactory(), chKafkaProducer);
+            logMessageProcessed(message, ORDER_RECEIVED_TOPIC);
+        } catch (ServiceException ex){
+            logMessageProcessingFailureRecoverable(message, ORDER_RECEIVED_TOPIC, ORDER_RECEIVED_TOPIC_ERROR, ex);
+            republishMessageToTopic(orderReceivedUri, ORDER_RECEIVED_TOPIC, ORDER_RECEIVED_TOPIC_ERROR);
+        } catch (Exception x) {
+            logMessageProcessingFailureNonRecoverable(message, ORDER_RECEIVED_TOPIC, x);
+        }
     }
 
-    @KafkaListener(topics = ORDER_RECEIVED_TOPIC, groupId = ORDER_RECEIVED_GROUP,
-            autoStartup = "#{!${uk.gov.companieshouse.item-handler.error-consumer}}")
-    public void processOrderReceived(String orderReceivedUri)
-            throws SerializationException, ExecutionException, InterruptedException {
+    /**
+     * Retry (`-retry`) listener/consumer. Calls a `retryable` method to process received message.
+     * If the `retryable` processor is unsuccessful with a `retryable` error, after maximum numbers of attempts allowed,
+     * the message is published to `-error` topic for failover processing.
+     * @param message
+     * @throws SerializationException
+     * @throws ExecutionException
+     * @throws InterruptedException
+     */
+    @KafkaListener(id = ORDER_RECEIVED_GROUP_RETRY, groupId = ORDER_RECEIVED_GROUP_RETRY,
+                    topics = ORDER_RECEIVED_TOPIC_RETRY,
+                    autoStartup = "#{!${uk.gov.companieshouse.item-handler.error-consumer}}")
+    public void processOrderReceivedRetry(org.springframework.messaging.Message<OrderReceived> message)
+            throws InterruptedException, ExecutionException, SerializationException {
+        OrderReceived orderReceived = message.getPayload();
```

Comment:
This is doing the same thing as the previous method (processOrderReceived). The logic should not be duplicated as this will risk bugs in the future if someone doesn't realise the retry method exists.

Reply:
With @KafkaListener annotation-based consumer/listener, each consumer can have it's own method, which improves for clarity. The alternative is a single method with a lot of if/else condition checking when processing and delegating the incoming message. 
All common processing is factored out to their own methods so there is no duplication.

Reply:
Can each listener permutation to have its own method as you say, but still share the largely same implementation so that any change made to the algorithm going forward will perforce cover all three case?

Reply:
Yes. Based on @jsimons1989's suggestion, all common operations (message processing and exception handling) have now been extracted to `handleMessage()` method. This helps separate message receiving from its processing. Any future consumer/listener performance related configuration changes can also be best applied individually to each consumer. For example, if the main topic needs to have separate performance needs.

The implementation for `main` and `-retry` topic can be combined with checks placed on received messages for differentiating which topic they came from, however apart from adding complexity to the current logic that approach would limit the possibility of future topic-specific configuration to do with performance etc.

The `-error` topic has a very specific function, in that it is enabled in an independently running app instance, and must pause consuming after reaching a desired `offset`. This requires that the `-error` topic listener be maintained as a separate consumer.